### PR TITLE
Listdir

### LINKED ISSUE #185: Expose fs.listdir
Currently, you can do 

```python
>>> api._authorities['osdc'].fs.listdir('/ACP/')
['climate', 'integration']
```

We could easily expose this by parsing  the directory as the authority_name, e.g.:

```python
>>> api.listdir('osdc://')
['climate', 'integration']
```

### PR DESCRIPTION
**Pull request by delgadom**
- [x] closes #185 
- [x] tests added / passed
- [x] passes ``git diff upstream/master | flake8 --diff``
- [x] whatsnew entry

In this PR:

* Implement listdir for DataAPI and cli (GH #185)
* Add whatsnew (GH #201)

Usage:

```python
>>> api.listdir('osdc://ACP/climate/')
['gcm-modelweights.csv', 'hddcdd', 'smme']
```

```bash
$ datafs listdir osdc://ACP/climate/
gcm-modelweights.csv
hddcdd
smme
```

I'm up for pushback on this one @jgerardsimcock and @jrising. There are a couple issues with this one:

* No good way to prevent listdir('valid_archive_name'), in which case we return a list of version numbers (the actual files). e.g.:
  ```python
  >>> api.listdir('osdc://ACP/climate/gcm-modelweights.csv')
  ['1.0', '1.1']
  ```
  The only way of preventing this would be to issue a call to the manager for each component returned by listdir to make sure they are not valid archive names, but this would kill the performance
* Since I rely on fs's `listdir`, we can only do this by authority_name, so the `osdc://` prefix (or the `authority_name` argument) is required unless `api.default_authority` returns without error (either there is only one authority or `DefaultAuthority` is set). This currently works for the GCP, but it will suddenly stop working when we add another authority (e.g. NASA's FTP server)

But, it was easy to implement and I think it yields significant usability benefits so I went ahead and created the PR

### CODE REVIEW

File: datafs/core/data_api.py [hunk 1]
```diff
@@ -169,6 +169,67 @@ def get_archive(self, archive_name, default_version=None):
...
+            location argument. This is because DataFS stores the archive path
+            as a directory and the versions as the actual files when versioning
+            is on.
+
+        Parameters
+        ----------
+
+        location : str
+
+            Path of the "directory" to search
```
Comment (jrising) [inline datafs/core/data_api.py]: Probably worthwhile to explain the convention for directories here.  "The virtual directory structure is based on the use of forward slashes (/) in the archive names."

Comment (delgadom) [inline datafs/core/data_api.py]: Sort of, although this reveals the aspect I'm a little squeamish about. `listdir` here is simply exposing the [listdir method](http://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.listdir) of the underlying Pyfilesystem `fs` object. So the path is interpreted by `fs`, not by `DataFS`. No convention about it. In our case, this is actually listing the virtual directory on OSDC's s3 storage.

I agree, though, that an article in the docs regarding archive naming conventions would be helpful.

**Comment by delgadom** [comment]
@jrising you can check for archive name validity using `DataAPI.get_archive()`. `is_archive` (doesn't currently exist) would have similar performance. Would that be useful?

**Comment by jrising** [comment]
It seems a little weird to be able to say `DataAPI.get_archive(...)`, and have that return successfully but with `is_archive` false.  What about just having a `DataAPI.is_archive(...)` call?  Or having `get_archive` throw an exception if it isn't an archive (which would be annoying to use as a filter with `listdir`, but semantically sensible.

**Comment by delgadom** [comment]
Sorry - I wasn't clear. `get_archive` raises a KeyError if the archive doesn't exist.

`is_archive` is not currently implemented, but could be without too much trouble if you think it would be useful.

**Comment by jrising** [comment]
I was still completely confused by this, and I finally figured out that it stemmed from reading your proposal above as making a function `get_archive().is_archive()`.  But that's not what you meant.

Yes, a `DataAPI.is_archive` function would be appreciated.

**Comment by delgadom** [comment]
Hi @jrising -

Thanks again for the feedback & thought. @jgerardsimcock  and I just talked about this and think that `DataAPI.is_archive()` would promote bad behavior. The reason why is because it would involve a separate call to the archive manager database, which is expensive from a database/latency standpoint, but returns very little information. Instead, you can simply wrap `get_archive` in a `try: ... except:` block for the same functionality.

Better still, only use listdir/search for exploring/finding archives, then use filter with glob-style patterns in production code so you're sure the archives exist and match what you're looking for.

If you can think of a use case that calls for `is_archive` that wouldn't be satisfied by `get_archive`, we'd be happy to implement it.

Also, I'm working on a batch version of `get_archive` (see #191) which will have dramatically better performance compared to multiple `get_archive` calls and should alleviate this problem slightly.

**Pull request merged**